nn.hpp

```diff
@@ -293,17 +293,21 @@ namespace std {
  */
 #include <stdexcept>
 
-// NN_CHECK_ASSERT triggers an assertion if expression is null.
-#define NN_CHECK_ASSERT(_e) (([&] (auto p) { \
+// NN_CHECK_ASSERT takes a pointer of type PT (e.g. raw pointer, std::shared_ptr or std::unique_ptr)
+// and returns a non-nullable pointer of type nn<PT>.
+// Triggers an assertion if expression evaluates to null.
+#define NN_CHECK_ASSERT(_e) (([&] (typename std::remove_reference<decltype(_e)>::type p) { \
         /* note: assert() alone is not sufficient here, because it might be compiled out. */ \
         assert(p && #_e " must not be null"); \
         if (!p) std::abort(); \
         return dropbox::oxygen::nn<typename std::remove_reference<decltype(p)>::type>( \
```

Comment:
The remove_reference applied to decltype(p) seems like it's probably redundant, now that the type of p is explicitly a value.  Doesn't hurt, though.